xds/internal/xdsclient/channel_test.go

```diff
@@ -0,0 +1,787 @@
+			opts:       xdsChannelOpts{transport: &fakeTransport{}},
+			wantErrStr: "serverConfig is nil",
+		},
+		{
+			name: "emptyBootstrapConfig",
+			opts: xdsChannelOpts{
+				transport:    &fakeTransport{},
+				serverConfig: &bootstrap.ServerConfig{},
+			},
+			wantErrStr: "bootstrapConfig is nil",
+		},
+		{
+			name: "emptyResourceTypeGetter",
+			opts: xdsChannelOpts{
+				transport:       &fakeTransport{},
+				serverConfig:    &bootstrap.ServerConfig{},
+				bootstrapConfig: &bootstrap.Config{},
+			},
+			wantErrStr: "resourceTypeGetter is nil",
+		},
+		{
+			name: "emptyEventHandler",
+			opts: xdsChannelOpts{
+				transport:          &fakeTransport{},
+				serverConfig:       &bootstrap.ServerConfig{},
+				bootstrapConfig:    &bootstrap.Config{},
+				resourceTypeGetter: func(string) xdsresource.Type { return nil },
+			},
+			wantErrStr: "eventHandler is nil",
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			_, err := newXDSChannel(test.opts)
+			if err == nil || !strings.Contains(err.Error(), test.wantErrStr) {
+				t.Fatalf("newXDSChannel() = %v, want %q", err, test.wantErrStr)
+			}
+		})
+	}
+}
+
+// Tests different scenarios of the xdsChannel receiving a response from the
+// management server. In all scenarios, the xdsChannel is expected to pass the
+// received responses as-is to the resource parsing functionality specified by
+// the resourceTypeGetter.
+func (s) TestChannel_ADS_HandleResponseFromManagementServer(t *testing.T) {
+	const (
+		listenerName1 = "listener-name-1"
+		listenerName2 = "listener-name-2"
+		routeName     = "route-name"
+		clusterName   = "cluster-name"
+	)
+	var (
+		badlyMarshaledResource = &anypb.Any{
+			TypeUrl: "type.googleapis.com/envoy.config.listener.v3.Listener",
+			Value:   []byte{1, 2, 3, 4},
+		}
+		apiListener = &v3listenerpb.ApiListener{
+			ApiListener: testutils.MarshalAny(t, &v3httppb.HttpConnectionManager{
+				RouteSpecifier: &v3httppb.HttpConnectionManager_RouteConfig{
+					RouteConfig: &v3routepb.RouteConfiguration{
+						Name: routeName,
+						VirtualHosts: []*v3routepb.VirtualHost{{
+							Domains: []string{listenerName1},
+							Routes: []*v3routepb.Route{{
+								Match: &v3routepb.RouteMatch{
+									PathSpecifier: &v3routepb.RouteMatch_Prefix{Prefix: "/"},
+								},
+								Action: &v3routepb.Route_Route{
+									Route: &v3routepb.RouteAction{
+										ClusterSpecifier: &v3routepb.RouteAction_Cluster{Cluster: clusterName},
+									}}}}}}},
+				},
+				HttpFilters: []*v3httppb.HttpFilter{e2e.RouterHTTPFilter},
+				CommonHttpProtocolOptions: &v3corepb.HttpProtocolOptions{
+					MaxStreamDuration: durationpb.New(time.Second),
+				},
+			}),
+		}
+		listener1 = testutils.MarshalAny(t, &v3listenerpb.Listener{
+			Name:        listenerName1,
+			ApiListener: apiListener,
+		})
+		listener2 = testutils.MarshalAny(t, &v3listenerpb.Listener{
+			Name:        listenerName2,
+			ApiListener: apiListener,
```

Comment:
how can we use same apiListener here for listener2? The `VirtualHosts`'s Domains have `listenerName1` only.